File: frontend/src/api/operations.ts
import client from './client';

export type AgentStatus = 'DRAFT' | 'ACTIVE' | 'DEPRECATED' | 'SUSPENDED';

export interface OperationsOverview {
  total_agents: number;
  status_breakdown: {
    ACTIVE: number;
    DRAFT: number;
    DEPRECATED: number;
    SUSPENDED: number;
  };
  visibility_breakdown: Record<string, number>;
  total_interactions: number;
  interactions_24h: number;
  failed_interactions: number;
  success_rate: number;
  total_tokens: number;
}

export interface AgentOps {
  agent_id: string;
  name: string;
  version: string;
  status: AgentStatus;
  visibility: string;
  owner_name: string | null;
  tool_count: number;
  updated_at: string;
  last_activity: string | null;
  active_executions: number;
  control_generation: number;
}

export interface Activity {
  interaction_id: string;
  actor_name: string;
  target_name: string | null;
  kind: string;
  state: string;
  start_timestamp: string;
  duration_ms: number | null;
  model_used: string | null;
  error_message: string | null;
}

export interface HealthComponent {
  name: string;
  status: 'online' | 'degraded' | 'offline';
  detail: string;
}

export interface ExecutionSummary {
  execution_tree_id: string;
  root_interaction_id: string;
  conversation_id: string;
  actor_name: string;
  prompt: string | null;
  state: string;
  node_count: number;
  duration_ms: number | null;
  started_at: string;
}

export interface ExecutionNode {
  interaction_id: string;
  parent_id: string | null;
  execution_tree_id: string;
  tree_depth: number;
  tree_path: string;
  actor_name: string;
  target_name: string | null;
  kind: string;
  state: string;
  duration_ms: number | null;
  reasoning_trace: string | null;
  results: string | null;
  tool_name: string | null;
  error_message: string | null;
  start_timestamp: string;
  payload: {
    schema_version: string;
    source_schema_version: string;
    input?: string | null;
    output?: string | null;
    reasoning?: string | null;
    tool?: { name?: string | null; arguments?: unknown; result?: unknown } | null;
    metadata?: Record<string, unknown>;
  };
}

export interface OperationsAnalytics {
  models: Array<{
    model: string;
    execution_count: number;
    failed_count: number;
    token_input: number;
    token_output: number;
    total_tokens: number;
    average_duration_ms: number;
    estimated_cost_usd: number | null;
  }>;
  parallel_groups: Array<{
    parallel_group_id: string;
    execution_count: number;
    wall_duration_ms: number;
    serial_duration_ms: number;
    saved_duration_ms: number;
  }>;
}

export interface ConnectorStatus {
  configured: boolean;
  endpoint: string | null;
  production_https_required: boolean;
}

export const operationsApi = {
  getOverview: () =>
    client.get<OperationsOverview>('/operations/overview').then((r) => r.data),
  getAgents: () => client.get<AgentOps[]>('/operations/agents').then((r) => r.data),
  updateAgentStatus: (agentId: string, status: AgentStatus) =>
    client
      .patch<AgentOps>(`/operations/agents/${agentId}/status`, { status })
      .then((r) => r.data),
  getActivity: (limit = 25) =>
    client
      .get<Activity[]>('/operations/activity', { params: { limit } })
      .then((r) => r.data),
  getHealth: () =>
    client
      .get<{ components: HealthComponent[] }>('/operations/health')
      .then((r) => r.data.components),
  getExecutions: (limit = 20) =>
    client.get<ExecutionSummary[]>('/operations/executions', { params: { limit } }).then((r) => r.data),
  getExecutionTree: (executionTreeId: string) =>
    client.get<{ execution_tree_id: string; nodes: ExecutionNode[] }>(`/operations/executions/${executionTreeId}`).then((r) => r.data),
  getAnalytics: () =>
    client.get<OperationsAnalytics>('/operations/analytics').then((r) => r.data),
  getSecurityConnector: () =>
    client.get<ConnectorStatus>('/operations/connectors/security-webhook').then((r) => r.data),
  testSecurityConnector: () =>
    client.post<{ configured: boolean; delivered: boolean; status_code: number | null; error?: string }>('/operations/connectors/security-webhook/test').then((r) => r.data),
};

File: backend/app/schemas/operations.py
"""MeshBoard — 운영 관리(Operations Console) 스키마."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Dict, List, Optional

from pydantic import BaseModel, Field


# ── Overview ──────────────────────────────────────────────────────
class AgentStatusBreakdown(BaseModel):
    ACTIVE: int = 0
    DRAFT: int = 0
    DEPRECATED: int = 0
    SUSPENDED: int = 0


class OperationsOverview(BaseModel):
    total_agents: int
    status_breakdown: AgentStatusBreakdown
    visibility_breakdown: Dict[str, int] = Field(default_factory=dict)
    total_interactions: int
    interactions_24h: int
    failed_interactions: int
    success_rate: float  # 0~100 (%)
    total_tokens: int


# ── Agent Lifecycle ───────────────────────────────────────────────
class AgentOpsRead(BaseModel):
    agent_id: uuid.UUID
    name: str
    version: str
    status: str
    visibility: str
    owner_name: Optional[str] = None
    tool_count: int = 0
    updated_at: datetime
    last_activity: Optional[datetime] = None
    active_executions: int = 0
    control_generation: int = 0


class AgentStatusUpdate(BaseModel):
    status: str = Field(..., max_length=20)


# ── Activity Log ──────────────────────────────────────────────────
class ActivityRead(BaseModel):
    interaction_id: uuid.UUID
    actor_name: str
    target_name: Optional[str] = None
    kind: str
    state: str
    start_timestamp: datetime
    duration_ms: Optional[int] = None
    model_used: Optional[str] = None
    error_message: Optional[str] = None


class ExecutionSummaryRead(BaseModel):
    execution_tree_id: uuid.UUID
    root_interaction_id: uuid.UUID
    conversation_id: uuid.UUID
    actor_name: str
    prompt: Optional[str] = None
    state: str
    node_count: int
    duration_ms: Optional[int] = None
    started_at: datetime


class ExecutionNodeRead(BaseModel):
    interaction_id: uuid.UUID
    parent_id: Optional[uuid.UUID] = None
    execution_tree_id: uuid.UUID
    tree_depth: int
    tree_path: str
    actor_name: str
    target_name: Optional[str] = None
    kind: str
    state: str
    duration_ms: Optional[int] = None
    reasoning_trace: Optional[str] = None
    results: Optional[str] = None
    tool_name: Optional[str] = None
    error_message: Optional[str] = None
    start_timestamp: datetime
    payload: Dict[str, object] = Field(default_factory=dict)


class ExecutionTreeRead(BaseModel):
    execution_tree_id: uuid.UUID
    nodes: List[ExecutionNodeRead] = Field(default_factory=list)


class ArchiveResultRead(BaseModel):
    retention_days: int
    cutoff: datetime
    eligible_count: int
    archived_count: int
    dry_run: bool


class ModelAnalyticsRead(BaseModel):
    model: str
    execution_count: int
    failed_count: int
    token_input: int
    token_output: int
    total_tokens: int
    average_duration_ms: float
    estimated_cost_usd: Optional[float] = None


class ParallelGroupAnalyticsRead(BaseModel):
    parallel_group_id: uuid.UUID
    execution_count: int
    wall_duration_ms: int
    serial_duration_ms: int
    saved_duration_ms: int


class OperationsAnalyticsRead(BaseModel):
    models: List[ModelAnalyticsRead] = Field(default_factory=list)
    parallel_groups: List[ParallelGroupAnalyticsRead] = Field(default_factory=list)


class ConnectorStatusRead(BaseModel):
    configured: bool
    endpoint: Optional[str] = None
    production_https_required: bool = True


class ConnectorTestRead(BaseModel):
    configured: bool
    delivered: bool
    status_code: Optional[int] = None
    error: Optional[str] = None


# ── System Health ─────────────────────────────────────────────────
class HealthComponent(BaseModel):
    name: str
    status: str  # "online" | "degraded" | "offline"
    detail: str


class SystemHealth(BaseModel):
    components: List[HealthComponent]
